core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/bavet/BavetConstraintSessionFactory.java

```diff
@@ -57,7 +57,8 @@ public BavetConstraintSessionFactory(SolutionDescriptor<Solution_> solutionDescr
                 constraintMatchEnabled);
 
         Score_ zeroScore = scoreDefinition.getZeroScore();
-        Set<BavetAbstractConstraintStream<Solution_>> constraintStreamSet = new LinkedHashSet<>(constraintList.size() * 10);
+        // Node sharing already happened, so we no longer need to use stream equality; stream identity will be enough.
+        Set<BavetAbstractConstraintStream<Solution_>> constraintStreamSet = new ActiveStreamSupport<>();
```

Comment:
Btw, we might be able to do `new LinkedIdentityHashMap().entrySet()`

Reply:
There would not. But I wrote this defensively - someone at some point may screw up `equals(...)` in  a stream, and it will lead to weird failures in this seemingly unrelated code.
If we use identity here, that becomes impossible. Equality bugs will be isolated to node sharing.